### PR TITLE
Fail loudly when running TorchAgent in mturk if interactive mode is False

### PR DESCRIPTION
**Pull request by stephenroller**
**Patch description**
If we create an mturk task, the shared TorchAgents need to ensure they don't share `self.replies`, otherwise different workers can trample all over each other. This was addressed in #1287, and worked around by not sharing replies when `batchsize == 1` .

However, it's still very easy to forget to set this work around! This patch attempts to make it so that TorchAgent would've loudly complained and crashed. This is done by:

1. Having mturk set interactive mode by default, which makes the default sane and
2. Having an assert in `act()` which will fail if we're in the dangerous setting.

**Testing steps**
Added two tests in TestTorchAgent. Before fixing TorchAgent, they both fail with:
```
$ python tests/test_torch_agent.py
/private/home/roller/working/parlai/parlai/core/metrics.py:38: UserWarning: Rouge metrics require py-rouge. Please run `pip install py-rouge`.
  warn_once('Rouge metrics require py-rouge. Please run `pip install py-rouge`.')
......F...F...
======================================================================
FAIL: test_interactive_mode (__main__.TestTorchAgent)
Test if conversation history is destroyed in MTurk mode.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/test_torch_agent.py", line 977, in test_interactive_mode
    response = shared.act()
AssertionError: RuntimeError not raised

======================================================================
FAIL: test_mturk_racehistory (__main__.TestTorchAgent)
Emulate a setting where batch_act misappropriately handles mturk.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/test_torch_agent.py", line 1000, in test_mturk_racehistory
    self.assertNotIn('thread2-msg1', share1.history.get_history_str())
AssertionError: 'thread2-msg1' unexpectedly found in 'thread1-msg1\nEvaluating 0 (responding to thread2-msg1)!\nthread1-msg2\nEvaluating 0 (responding to thread1-msg2)!\nthread1-msg3'

----------------------------------------------------------------------
Ran 14 tests in 0.117s
```

After patching TorchAgent:
```
$ python tests/test_torch_agent.py
/private/home/roller/working/parlai/parlai/core/metrics.py:38: UserWarning: Rouge metrics require py-rouge. Please run `pip install py-rouge`.
  warn_once('Rouge metrics require py-rouge. Please run `pip install py-rouge`.')
..............
----------------------------------------------------------------------
Ran 14 tests in 0.171s

OK
```

### CODE REVIEW

**Comment by stephenroller** [comment]
Wow, what a merge.

```
$ python tests/test_torch_agent.py
/private/home/roller/working/parlai/parlai/core/metrics.py:38: UserWarning: Rouge metrics require py-rouge. Please run `pip install py-rouge`.
  warn_once('Rouge metrics require py-rouge. Please run `pip install py-rouge`.')
...............
----------------------------------------------------------------------
Ran 15 tests in 3.850s

OK
```

And if I undo the changes to TA:

```
/private/home/roller/working/parlai/parlai/core/metrics.py:38: UserWarning: Rouge metrics require py-rouge. Please run `pip install py-rouge`.
  warn_once('Rouge metrics require py-rouge. Please run `pip install py-rouge`.')
......F...F....
======================================================================
FAIL: test_interactive_mode (__main__.TestTorchAgent)
Test if conversation history is destroyed in MTurk mode.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/test_torch_agent.py", line 903, in test_interactive_mode
    response = shared.act()
AssertionError: RuntimeError not raised

======================================================================
FAIL: test_mturk_racehistory (__main__.TestTorchAgent)
Emulate a setting where batch_act misappropriately handles mturk.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/test_torch_agent.py", line 925, in test_mturk_racehistory
    self.assertNotIn('thread2-msg1', share1.history.get_history_str())
AssertionError: 'thread2-msg1' unexpectedly found in 'thread1-msg1\nEvaluating 0 (responding to thread2-msg1)!\nthread1-msg2\nEvaluating 0 (responding to thread1-msg2)!\nthread1-msg3'

----------------------------------------------------------------------
Ran 15 tests in 3.977s

FAILED (failures=2)
```

File: parlai/core/torch_agent.py [hunk 1]
```diff
@@ -1671,6 +1674,11 @@ def reset_metrics(self):
 
     def act(self):
         """Call batch_act with the singleton batch."""
+        if self._replies_are_shared:
```
Comment (emilydinan) [inline parlai/core/torch_agent.py]: why are we raising a runtime error here??? 

am i misunderstanding something, or will this raise a runtime error every time we call act on an interactive agent

File: parlai/core/torch_agent.py [hunk 2]
```diff
@@ -1723,8 +1731,9 @@ def eval_step(self, batch):
 
     def set_interactive_mode(self, mode, shared):
         """Set interactive mode on or off."""
-        # Base class is a no-op.
-        pass
+        if shared is None:
```
Comment (emilydinan) [inline parlai/core/torch_agent.py]: do TRA and TGA call super() ? or just override this

Comment (emilydinan) [reply parlai/core/torch_agent.py]: ahh...nvm i see the answer to my question below 😄

File: parlai/core/torch_agent.py (hunk 1 above)
Comment (emilydinan) [reply parlai/core/torch_agent.py]: nevermind, ignore this comment lol

File: parlai/core/params.py [hunk 3]
```diff
@@ -432,6 +432,11 @@ def add_mturk_args(self):
             help='Specify location to use for scratch builds and such.',
         )
 
+        # it helps to indicaste to agents that they're in interactive mode, and
+        # can avoid some otherwise troublesome behavior (not having candidates,
+        # sharing self.replies, etc).
+        self.set_defaults(interactive_mode=True)
+
```
[the remark this answers is not in the document]
Comment (stephenroller) [reply parlai/core/params.py]: I don't think it matters, but changed for clarity and consistency.

**Pull request merged**